### PR TITLE
 [JSON-RPC]Audio library Fixes

### PR DESCRIPTION
**Pull request by DaveTBlake**
Fixes in one **breaking change** of  JSON_RPC to a number of long standing faults in the API access to the music library. Bumping major version to **9** as this is the first such API change for v18.

Since the API is already broken in these places one could say that anything done fixing it is not a breaking change.  However the way it is broken is well disguised, and leaving deprecated fields very misleading, this seems a good point to remove them fully and cleanly.

@Montellese could you review this please, I really do want to get music working better via the API.

~~This PR depends upon #13042 which needs to merge first (and is the first commit here)~~

Changes in more detail:

1) **GetAlbums() and GetAlbumDetails()**
While the inconsistencies in album genre handling within Kodi have are fixed by #13042 , that does not change the data exposed by the API. That data that is exposed is still not the matching and consistent data API consumers expect it to be.  `GetAlbums()` and `GetAlbumDetails()` both return "genre" (string array) and "genreid" (int array) fields, with the _original intention/belief_  that they match. But "genre" comes from `Album.strGenres` (and could be anything, although often is song genres) and "genreid" is the distinct (song) genre ids for all the songs on the album (from the `genre` table via `album_genre`) . 
Hence API changes: 
- **Remove** "genreid"  from `Audio.Details.Media`
- Add "genreid" to `Audio.Details.Song`
- Add "songgenres", an array of {title, id} pairs of matching song genre strings and id, to `Audio.Fields.Album`  and  `Audio.Details.Album`

So that "genreid"  is no longer returned by `GetAlbums()` and `GetAlbumDetails()`, but still returned by `GetSongs()` where it matches the song genre values returned in "genre".  And "songgenres", is returned by `GetAlbums()` and `GetAlbumDetails()` with matching song genre strings and id pairs.

That approach matches how `GetArtists()` and `GetArtistDetails()` works. 
For artists "genre" is artist genres, for albums it is album genres and they both have "songgenres" field, for songs "genre" is song genres and songs also have a "genreid" field. This field naming is also consistent with how "artist" is handled: in album "artist" is album artists, in song it is song artists and there is an "albumartist" field. But note that `GetAlbums()`  filters "genreid" or "genre" are song genre id or string respectively.

We want to alert all API consumers to use "songgenres" - a matching pairs of  string and id - if they want genre values that enable them to replicate Kodi genre filtering. I think it best to remove "genreid" from fields returned for albums, and have calls with "genreid" fail rather than risk API consumers continue to it in a mistaken way. But also make clear what is in "genre", for albums this continues as before to return album genre strings (that may be anything not necessarily song genres).

2) **SetAlbumDetails() and SetSongDetails()**
This PR fixes the long outstanding bugs in `SetSongDetails()`  https://trac.kodi.tv/ticket/15247. 
Since Gotham and the introduction of artist credits and Musicbrainz ids,  neither `SetSongDetails()` nor `SetAlbumDetails()` was modified sufficiently for them to work fully. In particular updating artists or album artists is broken, see https://forum.kodi.tv/showthread.php?tid=195711. `SetSongDetails()` also has album fields that should be part of  `SetAlbumDetails()` instead.
Hence API changes: 
- Add parameters "musicbrainzalbumartistid" (string array) and "displayartist" to `SetAlbumDetails()`
- Remove album properties: "albumartist", "album", "musicbrainzalbumid", "musicbrainzalbumartistid" from `SetSongDetails()`
- Add string parameters "displayartist", "sortartist", "mood" to `SetSongDetails()`

3) **Other Deprecated Fields and Inconsistencies**
Since we are making a breaking change in this area we can also drop other deprecated fields, particularly not leaving fields that only apply to albums returned blank for songs (see #12963) and risking further confusion.
Hence API changes: 
- Remove "musicbrainzalbumid" from `Audio.Details.Media` 
- Add "musicbrainzalbumid" to `Audio.Details.Albums`
- Correct "musicbrainzartistid" in `Audio.details.Song` , `Audio.details.Artist` and `List.Item.Base` to be an array. Although for artists it will always be an array of one element, it is always serialized consistently, and correctly as an array.

4) **"artist" Field Specification Conflict**
The above _consistent_ definition of "musicbrainzartistid" avoids the problem that exists with "artist".  JSON-RPC implicitly needs the "artist" field to always be the same type, but `Audio.Details.Media` defines it as an array, while `Audio.Details.Artist` defines it as a single string (logical as an artist is only going to have one name, but why is the name file of an artist called "artist" not "name"?).  Thus when it is serialized the mediatype is used to determine how to format - string or array. `List.Item.Base` extends `Audio.Details.Media` so expects an array, but when the fileitem list media type is "artists" a single string is returned (alla `Audio.Details.Artist` definition). 

This means for example that `Files.GetDirectory()` on an artists type smartplaylist returns "artist" as a string, breaking the specification given in `List.Item.Base`.  

However the use of the "artist" field is so core that I am not confident that making the spec consistent (say by making "artist" always an array), is worth the disruption.
Hence **no changes made, but inconsistency noted**.

~~I also think that  "albumreleasetype" should be removed from `Audio.Fields.Songs`. Releasetype is an _internal_ abum property (with values either "album" or "single") indicating if the album is a real or a fake album created for colating the singles by an artists. It is not populated when songs are quieried from the library, and only set in CMusicInfoTag (hence serialized) when mediatype is album.~~

### CODE REVIEW

**Comment by DaveTBlake** [comment]
@SyncedSynapse and @joethefox this is of interest to you, especially the album genre parts

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 1]
```diff
@@ -580,19 +580,35 @@ JSONRPC_STATUS CAudioLibrary::SetAlbumDetails(const std::string &method, ITransp
...
-  // As temp fix just set the album artist description string
+  if (ParameterNotNull(parameterObject, "displayartist")) 
+    album.strArtistDesc = parameterObject["displayartist"].asString();
+  // Set album sort string before processing artist credits
+  if (ParameterNotNull(parameterObject, "sortartist"))
+    album.strArtistSort = parameterObject["sortartist"].asString();
+
+  // Match up artist names and mbids to make new artist credits
+  // Mbid values only apply if there are names
+  bool updateartists = false;
```
Comment (Rechi) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp]: unused but set variable

**Comment by DaveTBlake** [comment]
LINUX-64-Wayland fail is unrelated to this PR, rest builds fine

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 2]
```diff
@@ -580,19 +580,35 @@ JSONRPC_STATUS CAudioLibrary::SetAlbumDetails(const std::string &method, ITransp
...
+    album.strArtistSort = parameterObject["sortartist"].asString();
+
+  // Match up artist names and mbids to make new artist credits
+  // Mbid values only apply if there are names
   if (ParameterNotNull(parameterObject, "artist"))
   {
-    std::vector<std::string> artist;
-    CopyStringArray(parameterObject["artist"], artist);
-    album.strArtistDesc = StringUtils::Join(artist, g_advancedSettings.m_musicItemSeparator);
+    std::vector<std::string> artists, mbids;
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp]: Could you explicitly declare both variables on separate lines? I always find that declaring multiple variables on the same line results in people missing all but the first (or the last) variable name.

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 3]
```diff
@@ -908,29 +930,19 @@ JSONRPC_STATUS CAudioLibrary::GetAdditionalAlbumDetails(const CVariant &paramete
...
   std::set<std::string> additionalProperties;
   if (!CheckForAdditionalProperties(parameterObject["properties"], checkProperties, additionalProperties))
     return OK;
 
-  for (int i = 0; i < items.Size(); i++)
+  if (additionalProperties.find("songgenres") != additionalProperties.end())
   {
-    CFileItemPtr item = items[i];
-    if (additionalProperties.find("genreid") != additionalProperties.end())
+    for (int i = 0; i < items.Size(); i++)
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp]: Isn't `CFileItemList` iterable now?

**Review by Montellese — CHANGES_REQUESTED** [review]
First of all thanks for taking the time to update the JSON-RPC API and trying to get it cleaned up. Much appreciated!

Looks good in general apart from two minors.

Two more things:
* Would it make sense to merge the remaining "genre" and "genreid" into a "genre" proprty of type `Audio.Details.Genres` while we are at it?
* I didn't know about "albumreleasetype" so if it's only for internal use and not even properly populated we should remove it from the API.

**Comment by DaveTBlake** [comment]
@Montellese thanks for finding time to review.

> Would it make sense to merge the remaining "genre" and "genreid" into a "genre" property of type Audio.Details.Genres while we are at it?

I don't think so, it would unnecessarily slow API data access. The only remaining type with "genreid" and "genre" separately is Audio.Details.Song. Fetching the "genre" data for songs does not take any extra processing time, but fetching "genreid" does, and only some API consumers want genre id values. 

But perhaps it would be better to replace the "genreid" field in Audio.Details.Song with "songgenres" of type Audio.Details.Genres, so ids always come with their related string as a pair. However keeping the separate "genre" field (for efficient data access)  means that  `GetSongs` and `GetSongsDetails` would return song genre strings in both "genre" and "songgenres" fields. Is that data duplication acceptable?

> I didn't know about "albumreleasetype" so if it's only for internal use and not even properly populated we should remove it from the API.
 
Will do.

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp (hunk 3 above)
Comment (DaveTBlake) [reply xbmc/interfaces/json-rpc/AudioLibrary.cpp]: Yes it is, but I would prefer to leave this loop in the same style as all the other item loops in this unit. To me that makes it more readable

**Comment by DaveTBlake** [comment]
On looking closer it seems I was mistaken about "albumreleasetype". It does get populated but by `CMusicDatabase::GetFileItemFromDataset()` rather than `CMusicInfoTag::SetSong()`, and correctly indicates if the song is a single or from an album. So nothing to change.

**Review by Montellese — APPROVED** [review]
> I don't think so, it would unnecessarily slow API data access. The only remaining type with "genreid" and "genre" separately is Audio.Details.Song. Fetching the "genre" data for songs does not take any extra processing time, but fetching "genreid" does, and only some API consumers want genre id values.

Right I didn't know that it required extra database queries.

> But perhaps it would be better to replace the "genreid" field in Audio.Details.Song with "songgenres" of type Audio.Details.Genres, so ids always come with their related string as a pair. However keeping the separate "genre" field (for efficient data access) means that GetSongs and GetSongsDetails would return song genre strings in both "genre" and "songgenres" fields. Is that data duplication acceptable?

I'm not a fan of data duplication in an API. Would have been nice to be consistent but if it means slowing responses down considerably that's not an option IMO.

Looks good to me, thanks.

**Comment by markusloffler** [comment]
Hi all, I just adapted my app to request the "songgenres" property for "AudioLibrary.GetAlbums".
"songgenres" is properly accepted when I do the JSON call (and "genreid" is not accepted anymore), but the resulting array does not contain any "songgenres" field. What could be wrong?

**Pull request merged**